### PR TITLE
Rust wrappers for hash and merkle API

### PR DESCRIPTION
**Pull request by yshekel**

### CODE REVIEW

File: wrappers/rust/icicle-core/src/hash/mod.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    pub fn hash<T>(
+        &self,
+        input: &(impl HostOrDeviceSlice<T> + ?Sized),
+        cfg: &HashConfig,
+        output: &mut (impl HostOrDeviceSlice<T> + ?Sized),
+    ) -> Result<(), eIcicleError> {
+        // check device slices are on active device
+        if input.is_on_device() && !input.is_on_active_device() {
+            eprintln!("input not allocated on an inactive device");
```
Comment (emirsoyturk) [inline wrappers/rust/icicle-core/src/hash/mod.rs:69]: eprintln!("input not allocated on an active device");
or 
eprintln!("input allocated on an inactive device");

File: wrappers/rust/icicle-core/src/hash/mod.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            return Err(eIcicleError::InvalidPointer);
+        }
+        if output.is_on_device() && !output.is_on_active_device() {
+            eprintln!("output not allocated on an inactive device");
```
Comment (emirsoyturk) [inline wrappers/rust/icicle-core/src/hash/mod.rs:73]: `eprintln!("output not allocated on an active device");`
or 
`eprintln!("output allocated on an inactive device");`

File: wrappers/rust/icicle-core/src/hash/mod.rs (hunk 1 above)
Comment (yshekel) [reply wrappers/rust/icicle-core/src/hash/mod.rs:69]: good catch. fixing

File: icicle/backend/cpu/src/hash/cpu_keccak.cpp [hunk 3]
```diff
@@ -6,15 +6,19 @@ namespace icicle {
...
     {
     }
 
     eIcicleError hash(const std::byte* input, uint64_t size, const HashConfig& config, std::byte* output) const override
     {
-      ICICLE_LOG_INFO << "Keccak CPU hash() called";
-      // TODO implement
+      ICICLE_LOG_DEBUG << "Keccak/sha3 CPU hash() called, batch=" << config.batch
+                       << ", single_output_size=" << output_size();
+      // TODO implement real logic
```
Comment (mickeyasa) [inline icicle/backend/cpu/src/hash/cpu_keccak.cpp:18]: prefer to add asserion

**Review by mickeyasa — APPROVED** [review]

File: wrappers/rust/icicle-core/src/merkle/mod.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+use std::{ffi::c_void, mem, ptr, slice};
+
+/// Enum representing the padding policy when the input is smaller than expected by the tree structure.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PaddingPolicy {
+    None,        // No padding, assume input is correctly sized.
+    ZeroPadding, // Pad the input with zeroes to fit the expected input size.
+    LastValue,   // Pad the input by repeating the last value.
+}
```
[the remark this answers is not in the document]
Comment (LeonHibnik) [reply wrappers/rust/icicle-core/src/merkle/mod.rs:14]: Probably yes but we will have to implement them separately

File: icicle/backend/cpu/src/hash/cpu_keccak.cpp (hunk 3 above)
Comment (yshekel) [reply icicle/backend/cpu/src/hash/cpu_keccak.cpp:18]: yes it's just that it will fail in CI. Aviad has a PR for that already.

**Pull request merged**